### PR TITLE
[wallet] [refactor] Simplify getbalance implementation

### PR DESCRIPTION
**Pull request by ryanofsky**
This is just code cleanup, no behavior is changing. It deduplicates getbalance code and makes it comprehsible so it is not doing bizarre things like subtracting negative fees (https://github.com/bitcoin/bitcoin/pull/9167).

For each wallet transaction, the balance is updated as follows:

```
+        CAmount debit = wtx.GetDebit(filter);
+        const bool outgoing = debit > 0;
+        for (const CTxOut& out : wtx.tx->vout) {
+            if (outgoing && IsChange(out)) {
+                debit -= out.nValue;
+            } else if (IsMine(out) & filter && depth >= minDepth && (!account || *account == GetAccountName(out.scriptPubKey))) {
+                balance += out.nValue;
+            }
+        }
+
+        if (outgoing && (!account || *account == wtx.strFromAccount)) {
+            balance -= debit;
+        }
```

### CODE REVIEW

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -687,6 +687,8 @@ UniValue getbalance(const JSONRPCRequest& request)
     if (request.params.size() == 0)
         return  ValueFromAmount(pwalletMain->GetBalance());
 
+    const string* account = request.params[0].get_str() != "*" ? &request.params[0].get_str() : nullptr;
```
Comment (jonasschnelli) [inline src/wallet/rpcwallet.cpp]: I think this results in no longer throwing `RPC_WALLET_INVALID_ACCOUNT_NAME` when someone tries to get balance of `"*"`, instead, it returns back the total balance. But I think this is what we want....

**Review by jonasschnelli — APPROVED** [review]
Nice.
utACK 4b49ed65ede769617eab048e4bbab6d5219600c6

I could imagine a follow-up refactoring that get rid of all the individual caches...
```
    mutable bool fDebitCached;
    mutable bool fCreditCached;
    mutable bool fImmatureCreditCached;
    mutable bool fAvailableCreditCached;
    mutable bool fWatchDebitCached;
    mutable bool fWatchCreditCached;
    mutable bool fImmatureWatchCreditCached;
    mutable bool fAvailableWatchCreditCached;
    mutable bool fChangeCached;
    mutable CAmount nDebitCached;
    mutable CAmount nCreditCached;
    mutable CAmount nImmatureCreditCached;
    mutable CAmount nAvailableCreditCached;
    mutable CAmount nWatchDebitCached;
    mutable CAmount nWatchCreditCached;
    mutable CAmount nImmatureWatchCreditCached;
    mutable CAmount nAvailableWatchCreditCached;
    mutable CAmount nChangeCached;
```
... and manages them in a `std::map` or similar.

File: src/wallet/rpcwallet.cpp (hunk 1 above)
Comment (ryanofsky) [inline src/wallet/rpcwallet.cpp]: Yes, this is just keeping the current behavior (line 698 before this commit, which is removed in the next commit).

**Review by ryanofsky — COMMENTED** [review]
> I could imagine a follow-up refactoring that get rid of all the individual caches...

I also think this could be good to unify/organize these. #9167 adds two more caches, so they could be integrated as well.

**Comment by ryanofsky** [comment]
Rebased 4b49ed65ede769617eab048e4bbab6d5219600c6 -> e8ce6a4877186940e95e67aeaa12fd11a23b1e0d ([getbalance-cleanup.0](https://github.com/ryanofsky/bitcoin/commits/pr/getbalance-cleanup.0) -> [getbalance-cleanup.1](https://github.com/ryanofsky/bitcoin/commits/pr/getbalance-cleanup.1)), fixing trivial merge conflict with #9613.

**Comment by ryanofsky** [comment]
Rebased e8ce6a4877186940e95e67aeaa12fd11a23b1e0d -> be74f34671ac89feaafe80b3ac43ff108c9ceb29 ([getbalance-cleanup.1](https://github.com/ryanofsky/bitcoin/commits/pr/getbalance-cleanup.1) -> [getbalance-cleanup.2](https://github.com/ryanofsky/bitcoin/commits/pr/getbalance-cleanup.2)) because of conflict with #9764.

**Comment by ryanofsky** [comment]
Rebased be74f34671ac89feaafe80b3ac43ff108c9ceb29 -> 0b9520d1fb242864b8bd3f6ec68fa91b231467df ([pr/getbalance-cleanup.2](https://github.com/ryanofsky/bitcoin/commits/pr/getbalance-cleanup.2) -> [pr/getbalance-cleanup.3](https://github.com/ryanofsky/bitcoin/commits/pr/getbalance-cleanup.3)) because of conflicts with pwalletMain renames in #9775.

**Comment by laanwj** [comment]
Needs rebase again (sorry!)

**Review by laanwj — APPROVED** [review]

**Comment by ryanofsky** [comment]
No problem, thanks for the review!

Rebased 0b9520d1fb242864b8bd3f6ec68fa91b231467df -> 02d9f50d5f3c96fe888c230d59c5afdab4c7c6a3 ([pr/getbalance-cleanup.3](https://github.com/ryanofsky/bitcoin/commits/pr/getbalance-cleanup.3) -> [pr/getbalance-cleanup.4](https://github.com/ryanofsky/bitcoin/commits/pr/getbalance-cleanup.4))

**Pull request merged**